api/agent/drivers/docker/imageCache.go

```diff
@@ -0,0 +1,223 @@
+func (c *Cache) contains(value d.APIImages) bool {
+	for _, i := range c.cache {
+		if i.image.ID == value.ID {
+			return true
+		}
+	}
+	return false
+
+}
+
+// Marks a image by ID as having been used.
+func (c *Cache) Mark(ID string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.mark(ID)
+}
+
+// Internal method for finding the image and adding to it's used count as well as it's lastUsed time.
+func (c *Cache) mark(ID string) error {
+	for idx, i := range c.cache {
+		if i.image.ID == ID {
+			c.cache[idx].lastUsed = time.Now()
+			c.cache[idx].uses = c.cache[idx].uses + 1
+			return nil
+		}
+	}
+
+	return errors.New("Image not found in cache")
+}
+
+// Remove deletes an image from the list. Also grabs a smaller portion of the slice.
+func (c *Cache) Remove(value d.APIImages) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for idx, i := range c.cache {
+		if i.image.ID == value.ID {
+			// Move the last item into the location of the item to be removed
+			c.cache[idx] = c.cache[len(c.cache)-1]
+			// shorten the list
+			c.cache = c.cache[:len(c.cache)-1]
+			return nil
+		}
+	}
+
+	return errors.New("Image not found in cache")
+}
+
+// Adds a token to the list of locks on this image by id.
+func (c *Cache) Lock(ID string, key interface{}) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.lock(ID, key)
+}
+
+// Internal method for Lock. Contains no locks.
+func (c *Cache) lock(ID string, key interface{}) error {
+	for _, i := range c.cache {
+		if i.image.ID == ID {
+			i.locked[&key] = &key
+			return nil
+		}
+	}
+	return errors.New("Image not found in cache")
+}
+
+// Check to see if an image by id as any locks on it.
+func (c *Cache) Locked(ID string) (bool, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.locked(ID)
+}
+
+func (c *Cache) locked(ID string) (bool, error) {
+	for _, i := range c.cache {
+		if i.image.ID == ID {
+			return len(i.locked) > 0, nil
+		}
+	}
+	return false, errors.New("Image not found in cache")
+}
+
+// Removes a token from an image by ID's lock set.
+func (c *Cache) Unlock(ID string, key interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.unlock(ID, key)
+}
+
+func (c *Cache) unlock(ID string, key interface{}) {
+	for _, i := range c.cache {
+		if i.image.ID == ID {
+			delete(i.locked, &key)
+		}
+	}
+}
+
+// Add adds the provided key and value to the cache, evicting
+// an old item if necessary.
+func (c *Cache) Add(value d.APIImages) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	logrus.Debugf("value: %v", value)
+	if c.contains(value) {
+		c.mark(value.ID)
+		return
+	}
+	c.cache = append(c.cache, NewEntry(value))
+}
+
+// Checks the total size of the docker data dir.
+func (c *Cache) TotalSize() int64 {
+	return 0
+}
+
+// Returns true if there is to much space consumed.
+func (c *Cache) OverFilled() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.TotalSize() > c.maxSize
+}
+
+// Checks for evictable images ordered by score.
+func (c *Cache) Evictable() EntryByAge {
```

Comment:
EntryByAge is a slice type, are there any restrictions on the length here? Should this be `[]Entry`?

Reply:
Its EntryByAre so it can be sorted.

Reply:
ok, i see what is going on here. Maybe the comment could be something like "Evictable returns all evictable image, ordered by score'